crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -153,303 +117,222 @@ impl<T: TransportConnect> Scheduler<T> {
-        );
-        self.ensure_leadership(&mut builder, placement_hints);
+        // todo(azmy): avoid cloning the partition table every time by keeping
+        // the latest built always available as a field
+        let mut builder = partition_table.clone().into_builder();
+        let replication_strategy = builder.replication_strategy();
 
-        if let Some(scheduling_plan) = builder.build_if_modified() {
-            let scheduling_plan = self
-                .try_update_scheduling_plan(scheduling_plan)
-                .await?
-                .into_inner();
+        builder.for_each(|partition_id, target_state| {
+            self.ensure_replication(
+                partition_id,
+                target_state,
+                alive_workers,
+                replication_strategy,
+                nodes_config,
+                &placement_hints,
+            );
+
+            self.ensure_leadership(partition_id, target_state, &placement_hints);
+        });
+
+        if let Some(partition_table) = builder.build_if_modified() {
+            debug!("Updated partition table placement: {partition_table:?}");
+            self.try_update_partition_table(version, partition_table)
+                .await?;
 
-            debug!("Updated scheduling plan: {scheduling_plan:?}");
-            self.assign_scheduling_plan(scheduling_plan);
+            return Ok(());
         }
 
         Ok(())
     }
 
-    fn assign_scheduling_plan(&mut self, scheduling_plan: SchedulingPlan) {
-        self.scheduling_plan = scheduling_plan;
-        self.last_updated_scheduling_plan = Instant::now();
-    }
-
-    async fn try_update_scheduling_plan(
+    async fn try_update_partition_table(
         &self,
-        scheduling_plan: SchedulingPlan,
-    ) -> Result<UpdateOutcome<SchedulingPlan>, Error> {
+        version: Version,
+        partition_table: PartitionTable,
+    ) -> Result<(), Error> {
         match self
             .metadata_store_client
             .put(
-                SCHEDULING_PLAN_KEY.clone(),
-                &scheduling_plan,
-                Precondition::MatchesVersion(self.scheduling_plan.version()),
+                PARTITION_TABLE_KEY.clone(),
+                &partition_table,
+                Precondition::MatchesVersion(version),
             )
             .await
         {
-            Ok(_) => Ok(UpdateOutcome::Written(scheduling_plan)),
-            Err(err) => match err {
-                WriteError::FailedPrecondition(_) => {
-                    // There was a concurrent modification of the scheduling plan. Fetch the latest version.
-                    let scheduling_plan = self.fetch_scheduling_plan().await?;
-                    Ok(UpdateOutcome::NewerVersionFound(scheduling_plan))
-                }
-                err => Err(err.into()),
-            },
+            Ok(_) => {}
+            Err(WriteError::FailedPrecondition(msg)) => {
+                debug!("Partition table update failed due to: {msg}");
```

Comment:
Should we try to sync the `PartitionTable` here in order to not wait for the regular update interval to happen? We do know that there must be a newer version available because of the `FailedPrecondition`.